### ISSUE TITLE
Matrix (and higher) comprehensions causes stackoverflow

### ISSUE THREAD

**Issue by zenna** [FluxML/Zygote.jl#176]
```julia
julia> x = rand(10, 10);

julia> gradient(x -> [i + j for i = 1:1.0:10, j = 1:1.0:10] .+ x, x)
ERROR: StackOverflowError:
Stacktrace:
 [1] _forward(::Zygote.Context, ::typeof(map), ::getfield(Main, Symbol("##8#10")), ::Base.Iterators.ProductIterator{Tuple{StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}},StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}) at /home/zenna/repos/Zygote.jl/src/compiler/interface2.jl:6
 [2] _forward(::Zygote.Context, ::typeof(collect), ::Base.Generator{Base.Iterators.ProductIterator{Tuple{StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}},StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}},getfield(Main, Symbol("##8#10"))}) at /home/zenna/repos/Zygote.jl/src/lib/array.jl:92
 [3] map at ./abstractarray.jl:2044 [inlined]
 [4] _forward(::Zygote.Context, ::typeof(map), ::getfield(Main, Symbol("##8#10")), ::Base.Iterators.ProductIterator{Tuple{StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}},StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}}) at /home/zenna/repos/Zygote.jl/src/compiler/interface2.jl:0
 ... (the last 3 lines are repeated 32476 more times)
 [97433] _forward(::Zygote.Context, ::typeof(collect), ::Base.Generator{Base.Iterators.ProductIterator{Tuple{StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}},StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}}},getfield(Main, Symbol("##8#10"))}) at /home/zenna/repos/Zygote.jl/src/lib/array.jl:92
 [97434] #7 at ./REPL[5]:1 [inlined]
 [97435] _forward(::Zygote.Context, ::getfield(Main, Symbol("##7#9")), ::Array{Float64,2}) at /home/zenna/repos/Zygote.jl/src/compiler/interface2.jl:0
 [97436] _forward(::Function, ::Array{Float64,2}) at /home/zenna/repos/Zygote.jl/src/compiler/interface.jl:31
 [97437] forward(::Function, ::Array{Float64,2}) at /home/zenna/repos/Zygote.jl/src/compiler/interface.jl:37
 [97438] gradient(::Function, ::Array{Float64,2}) at /home/zenna/repos/Zygote.jl/src/compiler/interface.jl:46
```

**Comment by MikeInnes** [comment]
Same as #154 I think. Our gradient for `collect` is bad right now.

**Comment by mtfishman** [comment]
Looks like this can be closed?
```julia
julia> using Zygote

julia> x = rand(10, 10);

julia> gradient(x -> sum(sum([i + j for i = 1:1.0:10, j = 1:1.0:10] .+ x)), x)
(Fill(1.0, 10, 10),)

(@v1.7) pkg> st Zygote
      Status `~/.julia/environments/v1.7/Project.toml`
  [e88e6eb3] Zygote v0.6.34
```

**Comment by darsnack** [comment]
Yes, I think this was closed in #785.
